controllers/utils.go

```diff
@@ -0,0 +1,135 @@
+// 2. It saves the manifests in the odh-manifests/component-name/ folder
+func downloadManifests(component odhdeploymentv1.Component) error {
+	// Get the component repo from the
+	resp, err := http.Get(component.URL)
+	if err != nil {
+		return fmt.Errorf("error downloading archive: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("error downloading archive: %v HTTP status", resp.StatusCode)
+	}
+
+	// Create a new gzip reader
+	gzipReader, err := gzip.NewReader(resp.Body)
+	if err != nil {
+		return fmt.Errorf("error creating gzip reader: %v", err)
+	}
+	defer gzipReader.Close()
+
+	// Create a new TAR reader
+	tarReader := tar.NewReader(gzipReader)
+
+	// Extract the contents of the TAR archive to the current directory
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		componentFiles := strings.Split(header.Name, "/")
+		componentFileName := ""
+		if len(componentFiles) >= 1 {
+			// Get the name and mode of the file/directory
+			componentFileName = strings.Join(componentFiles[1:], "/")
+		}
+		mode := header.FileInfo().Mode()
+
+		componentFoldersList := strings.Split(componentFileName, "/")
+		componentFolder := ""
+		relativeFilePath := ""
+		if len(componentFoldersList) >= 2 {
+			componentFolder = componentFoldersList[0]
+			relativeFilePath = strings.Join(componentFoldersList[1:], "/")
+		}
+
+		// If the entry is a directory, create it
+		if header.Typeflag == tar.TypeDir && componentFolder == component.ManifestFolder {
+
+			err = os.MkdirAll("odh-manifest/"+component.Name+"/"+relativeFilePath, mode)
+			if err != nil {
+				return fmt.Errorf("error creating directory:%v", err)
+			}
+			continue
+		}
+
+		// If the entry is a file, create it and download its contents
+		if header.Typeflag == tar.TypeReg && componentFolder == component.ManifestFolder {
+			fmt.Println("Creating file:", "odh-manifest/"+component.Name+"/"+relativeFilePath)
+			file, err := os.Create("odh-manifest/" + component.Name + "/" + relativeFilePath)
+			if err != nil {
+				fmt.Println("Error creating file:", err)
+			}
+			_, err = io.Copy(file, tarReader)
+			if err != nil {
+				fmt.Println("Error downloading file contents:", err)
+			}
+			file.Close()
+			continue
+		}
+	}
+
+	return err
+}
+
+func decodeK8sObjects(yamlData []byte, decoder runtime.Decoder) ([]*unstructured.Unstructured, error) {
+	var objs []*unstructured.Unstructured
+	reader := bytes.NewReader(yamlData)
+	infos, err := resource.NewBuilder(nil).Unstructured().Stream(reader, "").Flatten().Do().Infos()
+	if err != nil {
+		return objs, err
+	}
+
+	for _, info := range infos {
+		objs = append(objs, info.Object.(*unstructured.Unstructured))
+	}
+
+	return objs, nil
+}
+
+func (r *OdhDeploymentReconciler) createOrUpdate(ctx context.Context, instance *odhdeploymentv1.OdhDeployment, obj *unstructured.Unstructured) error {
```

Comment:
We can probably use https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.3/pkg/controller/controllerutil#CreateOrUpdate here